### PR TITLE
Stop words preprocessor and junits

### PR DESCRIPTION
**Pull request by vikasgupta78**
Stop words preprocessor and junits

### CODE REVIEW

**Review by sonalgoyal — CHANGES_REQUESTED** [review]

File: common/client/src/main/java/zingg/client/ZFrame.java [hunk 1]
```diff
@@ -1,9 +1,10 @@
 package zingg.client;
 
+import java.io.Serializable;
 import java.util.List;
 
 //Dataset, Row, column
-public interface ZFrame<D, R, C> {
+public interface ZFrame<D, R, C> extends Serializable{
```
[the remark this answers is not in the document]
Comment (vikasgupta78) [reply common/client/src/main/java/zingg/client/ZFrame.java]: fixed in commit a350297

File: common/core/src/main/java/zingg/preprocess/StopWords.java [hunk 2]
```diff
@@ -1,58 +1,110 @@
...
 import zingg.client.Arguments;
+import zingg.client.FieldDefinition;
 import zingg.client.ZFrame;
 import zingg.client.ZinggClientException;
 import zingg.client.util.ColName;
+import zingg.common.Context;
 import zingg.util.PipeUtilBase;
 
-public class StopWords<S,D,R,C,T> {
+public abstract class StopWords<S,D,R,C,T> implements Serializable{
```
[the remark this answers is not in the document]
Comment (vikasgupta78) [reply common/core/src/main/java/zingg/preprocess/StopWords.java]: done in commit db207e7 and e8220d4

File: common/core/src/main/java/zingg/preprocess/StopWords.java [hunk 3]
```diff
@@ -1,58 +1,110 @@
...
+		this.context = context;
+		this.args = args;
 	}
 
-	
-
-    public ZFrame<D,R,C> preprocessForStopWords(S session, Arguments args, ZFrame<D,R,C> ds) throws ZinggClientException {
-		/* 
+	public ZFrame<D, R, C> preprocessForStopWords(ZFrame<D, R, C> ds) throws ZinggClientException {
 		List<String> wordList = new ArrayList<String>();
```
[the remark this answers is not in the document]
Comment (vikasgupta78) [reply common/core/src/main/java/zingg/preprocess/StopWords.java]: done in d34fa3b

File: common/core/src/main/java/zingg/TrainingDataFinder.java [hunk 4]
```diff
@@ -57,7 +58,7 @@ public void execute() throws ZinggClientException {
 					
 				if (posPairs == null || posPairs.count() <= 5) {
 					ZFrame<D,R,C> posSamplesOriginal = getPositiveSamples(data);
-					ZFrame<D,R,C> posSamples = posSamplesOriginal ; //StopWords.preprocessForStopWords(args, posSamplesOriginal);
+					ZFrame<D,R,C> posSamples = getStopWords().preprocessForStopWords(posSamplesOriginal);
```
[the remark this answers is not in the document]
Comment (vikasgupta78) [reply common/core/src/main/java/zingg/TrainingDataFinder.java:61]: As discussed neg preprocess not needed as already done for training file

File: spark_zingg/core/src/main/java/zingg/spark/preprocess/SparkStopWords.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+	protected ZFrame<Dataset<Row>, Row, Column> removeStopWordsFromDF(ZFrame<Dataset<Row>, Row, Column> ds,
+			String fieldName, String pattern) {
+		
+		RemoveStopWordsUDF removeStopWordsUDF = new RemoveStopWordsUDF(pattern.toLowerCase());
+		// Each field will have different pattern
+		String udfName = removeStopWordsUDF.getName()+"_"+fieldName;
+		// register the UDF
+		getContext().getSession().udf().register(udfName, removeStopWordsUDF, DataTypes.StringType);
+		
+		Dataset<Row> dfAfterRemoval = ds.df().withColumn(fieldName,callUDF(udfName, ds.df().col(fieldName)));
```
[the remark this answers is not in the document]
Comment (vikasgupta78) [reply spark_zingg/core/src/main/java/zingg/spark/preprocess/SparkStopWords.java]: changed to udf2 in commit a46ccd9

File: common/core/src/main/java/zingg/preprocess/StopWords.java [hunk 6]
```diff
@@ -1,58 +1,110 @@
...
 import zingg.util.PipeUtilBase;
 
-public class StopWords<S,D,R,C,T> {
+public abstract class StopWords<S,D,R,C,T> implements Serializable{
 
+	private static final long serialVersionUID = 1L;
 	protected static String name = "zingg.preprocess.StopWords";
 	public static final Log LOG = LogFactory.getLog(StopWords.class);
-	protected static String stopWordColumn = ColName.COL_WORD;
+	protected String stopWordColumn = ColName.COL_WORD;
```
[the remark this answers is not in the document]
Comment (vikasgupta78) [reply common/core/src/main/java/zingg/preprocess/StopWords.java]: done in commit e2d8358

File: common/core/src/main/java/zingg/preprocess/StopWords.java [hunk 7]
```diff
@@ -1,58 +1,110 @@
...
-				//if (!Arrays.asList(stopWords.schema().fieldNames()).contains(stopWordColumn)) {
-					stopWordColumn = stopWords.columns()[COLUMN_INDEX_DEFAULT];
-				//}
-				wordList = stopWords.select(stopWordColumn).as(Encoders.STRING()).collectAsList();
-				String pattern = wordList.stream().collect(Collectors.joining("|", "\\b(", ")\\b\\s?"));
-				ds = ds.withColumn(def.getFieldName(), removeStopWords(pattern.toLowerCase()).apply(ds.col(def.getFieldName())));
+				ZFrame<D, R, C> stopWords = getStopWords(def);
+				resetStopWordColumn(stopWords);
+				wordList = getWordList(stopWords);
+				String pattern = getPattern(wordList);
```
[the remark this answers is not in the document]
Comment (vikasgupta78) [reply common/core/src/main/java/zingg/preprocess/StopWords.java]: done as part of e2d8358

**Pull request merged**